// @nozbe/watermelondb/utils/rx/__wmelonRxShimESM2015/index.js
"use strict";

exports.__esModule = true;
exports.startWith = exports.throttleTime = exports.switchMap = exports.map = exports.distinctUntilChanged = exports.multicast = exports.defer = exports.merge = exports.of = exports.BehaviorSubject = exports.ReplaySubject = exports.Subject = exports.Observable = void 0;

var _Observable = require("rxjs/_esm2015/internal/Observable");

exports.Observable = _Observable.Observable;

var _Subject = require("rxjs/_esm2015/internal/Subject");

exports.Subject = _Subject.Subject;

var _ReplaySubject = require("rxjs/_esm2015/internal/ReplaySubject");

exports.ReplaySubject = _ReplaySubject.ReplaySubject;

var _BehaviorSubject = require("rxjs/_esm2015/internal/BehaviorSubject");

exports.BehaviorSubject = _BehaviorSubject.BehaviorSubject;

var _of = require("rxjs/_esm2015/internal/observable/of");

exports.of = _of.of;

var _merge = require("rxjs/_esm2015/internal/observable/merge");

exports.merge = _merge.merge;

var _defer = require("rxjs/_esm2015/internal/observable/defer");

exports.defer = _defer.defer;

var _multicast = require("rxjs/_esm2015/internal/operators/multicast");

exports.multicast = _multicast.multicast;

var _distinctUntilChanged = require("rxjs/_esm2015/internal/operators/distinctUntilChanged");

exports.distinctUntilChanged = _distinctUntilChanged.distinctUntilChanged;

var _map = require("rxjs/_esm2015/internal/operators/map");

exports.map = _map.map;

var _switchMap = require("rxjs/_esm2015/internal/operators/switchMap");

exports.switchMap = _switchMap.switchMap;

var _throttleTime = require("rxjs/_esm2015/internal/operators/throttleTime");

exports.throttleTime = _throttleTime.throttleTime;

var _startWith = require("rxjs/_esm2015/internal/operators/startWith");

exports.startWith = _startWith.startWith;
